### ISSUE TITLE
Consensus hashes diverge and reconverge in Linux 32

### ISSUE THREAD

**Issue by zathras-crypto** [OmniLayer/omnicore#337]
MainNet Consensus comparing:

Linux x32
Linux x64
Win x32
Win x64
OSX x64

All match except Linux x32. which has the below diffs.

*Note Win x32 testing was done on WOW64 (32 bit binary on Windows 32-bit emulation in Win x64), will deploy a native x32 Win node for better testing.

Linux x32 has the following diffs:

``` diff
zathras@zathras-desktop:~/logs$ diff cut-linux64.log cut-linux32.log 
15581c15581
< Consensus hash for block 264906: cf2f73ed8bf37b8e2dd5c84d87e9cbe95baabc2ce1e2724dcb023f4639a39a21

---
> Consensus hash for block 264906: e7be3ace5f32a319645068237f3ba590d958f84d1054267f6dde049d04f61dbc
19618c19618
< Consensus hash for block 268695: 3612f8c673f36318eedcd0c444cd3cf0b7144d319a78797323ad5e5740f48565

---
> Consensus hash for block 268695: 1d3b6f97e8efd7b0c40ffcc5df5aa4a10d6c4c012bca0cf398c6d4a952571a53
21982,21983c21982,21983
< Consensus hash for block 270687: 4976d62701fa646fae62ff9ae6b5ed779bc63d4bc1cc87c287db26b2999b1945
< Consensus hash for block 270688: 4976d62701fa646fae62ff9ae6b5ed779bc63d4bc1cc87c287db26b2999b1945

---
> Consensus hash for block 270687: 14ffa229c30c447485b90f4b1e29a19b7bb802cad7893364616dd0da9885f79b
> Consensus hash for block 270688: 14ffa229c30c447485b90f4b1e29a19b7bb802cad7893364616dd0da9885f79b
26734c26734
< Consensus hash for block 274672: 79c6cfe14ec29fade768ef7bd5dfbf7b036f5fbee857a8a26e252670bade9c7e

---
> Consensus hash for block 274672: e07b7c74a15f2b2b86f2a0f9c5e5a547282bc0ceda5c11e94ff1e41cd48f1992
32329c32329
< Consensus hash for block 278854: fe56918fee01e8dbb8d83c3cd6424ccd7ae2e14aee26e39e9b8da52fb912ae09

---
> Consensus hash for block 278854: ea30ec329ba58e36c8f673430a0b0ea76d81af5df7f627615167f63a96e8c3f5
43882c43882
< Consensus hash for block 289061: 0214a2a4c5f423c6c5019ba7360da134c748314200396f96a17c47237543b6e5

---
> Consensus hash for block 289061: a23c850fb47aac7d01d1d8e598d2ca14f9fec35b1d0528c93040edeb5238396b
75803,75804c75803,75804
< Consensus hash for block 313440: 1a2b7e8ff778f61c21ea1ba8da099379a30fc75c4826617fdad925937acd6fdc
< Consensus hash for transaction c01273548866680ef71d64f3982a8cbbb5f64ea0a5945fa6766803cebec88360: e7dd59c33ded053477fc3e5999a04fbb39df3f4f86aedcceb24ded60abf1db6e

---
> Consensus hash for block 313440: 7333a42de5fd7edb2e8411f2f84dded65e856d1ec251a0548295890c7908fec3
> Consensus hash for transaction c01273548866680ef71d64f3982a8cbbb5f64ea0a5945fa6766803cebec88360: 4be3d644b379f0ad0244f8cdb5544687c7613a098d80b1e2930dd68329c299bd
91357c91357
< Consensus hash for block 325835: 49ef155de5d98d6037f35643df3c6ba79a6ac595722066dc5f98f51e8b19d1fe

---
> Consensus hash for block 325835: da33589e5dc9909374d73257aa23876938c45a1dc246cd356013e3f5ec1f4b1c
93293c93293
< Consensus hash for block 327456: 278c1592cbfadb9bb0c6479150d2a1405f97a5008dc7f3366f47c84b473932a9

---
> Consensus hash for block 327456: 30c12c845af152ea39b8f8fcc2c3fe8232c8b6648c2b23c4f0023695c2c5c191
95674c95674
< Consensus hash for block 329531: ae17e982846c9f222aece6bf0d729ed69e1e8c2db7adc1475d40381365720cc1

---
> Consensus hash for block 329531: dd8fe94e2839d1b1467228a8eacf4fe4ce3f1cf683a392c6973888b0351cce11
97681,97682c97681,97682
< Consensus hash for block 331179: 7938e5573ec224d6a141872717839a4974607421faa0434e3c80cff179b2f174
< Consensus hash for transaction c45fe4b6b5ecdbc84671ccdbd134c489c36059908706fa30600e1f603a1ae4dc: d50db9673f1d12fc2579c0fa2b561c9ebf911e6730c660dc69d986e7fcfd8d18

---
> Consensus hash for block 331179: f9c0b908198107f39d2c724a36442aaa7f6e32aa971ef97e66589bb72c3e20e0
> Consensus hash for transaction c45fe4b6b5ecdbc84671ccdbd134c489c36059908706fa30600e1f603a1ae4dc: 0e0d8c6caf83422f1d518c2cb3cd635b9f54454b6d7af480b9b6e6c90b5c1323
103581,103583c103581,103583
< Consensus hash for block 335829: 544f450305ef4477a32ad792fb6dbada1e86d00c279578b72c1a9691a0ad1db1
< Consensus hash for transaction 85bfa81fff23ca6ef30df685f4e7010316f3364400b4583c29b3bc64623e7ad6: c7287944247e99bd6d31a8687a2a4a600cbc8723497ef7d1b185099358732753
< Consensus hash for transaction c776e3822344b40fb77bc30f54e0e99754c54703c191295844490d80287da873: 312ef95fa0597577fe5bf96866500b935c20d27f46f8ecf0d9a514ce62b561b7

---
> Consensus hash for block 335829: dc21fd47af3de08b0ead7a4a61bdc334e0c656bb030634844a4695509cf1f1cb
> Consensus hash for transaction 85bfa81fff23ca6ef30df685f4e7010316f3364400b4583c29b3bc64623e7ad6: f6b54e4f67ebeb3f8dec62b2c80d8f689892722656a91b11eec73ef6fb25a21e
> Consensus hash for transaction c776e3822344b40fb77bc30f54e0e99754c54703c191295844490d80287da873: bcb84655d4b267cc5ea67560d774b92d65e50bfd009ad2526528ef2c1cfbe430
111804c111804
< Consensus hash for block 342628: f1fe3c753a13c7e8122e250b3069380102fbdbea7735c4b134b34fcaaa517a21

---
> Consensus hash for block 342628: f1254893f28079c6e409941c311b543f5bffcb5cc90c0f407413ebca788dd3fa
112241c112241
< Consensus hash for block 343009: fe75982888eca39800711432df1ae3f13e753f9894eebde941abfbe8c9643ded

---
> Consensus hash for block 343009: b336c76cade030e607de4b142bfd82fdc35efcd440ad88136ca1f15f9b932a0c
124537c124537
< Consensus hash for block 353292: 08d196d9e4159331abbc873f7e2e62075f34c6bca36917e8b18e6ed83a4b4b7d

---
> Consensus hash for block 353292: 1f292eb510c45e8decc92456c37ac07114b51a255eab2df2bef0da140e328c5f
129148,129151c129148,129151
< Consensus hash for block 356516: c1083a32b78a228380201ab97941d4509b30b35d7cb9c48ef1782289219f09e3
< Consensus hash for transaction 016255cefb551abedb4faf6de1488dcf0d12332d9764a9265c7d26fac2e02b81: d0f476df716cd64ab38c853425dda9b49ed6954dd090cd304579650aa3415484
< Consensus hash for block 356517: d0f476df716cd64ab38c853425dda9b49ed6954dd090cd304579650aa3415484
< Consensus hash for transaction 4167238249da6ccc07fc0bedc6865b91aba1d392af0f6839fd63f204a015f54c: a1e0cf63db6b2b987083842f5a3eafff16ab7e2eee92043e2a5d1b9e2fdfc0bb

---
> Consensus hash for block 356516: b9536b083cc7f9ad4735a8622c350d023ecd89de87e554e8a7391fcc75567358
> Consensus hash for transaction 016255cefb551abedb4faf6de1488dcf0d12332d9764a9265c7d26fac2e02b81: 33e794bc830b4fe29344b494985710685751e73d0dbbf858229b70e1533785b1
> Consensus hash for block 356517: 33e794bc830b4fe29344b494985710685751e73d0dbbf858229b70e1533785b1
> Consensus hash for transaction 4167238249da6ccc07fc0bedc6865b91aba1d392af0f6839fd63f204a015f54c: 1b96bd1ebd9ec730e7b6986bee4236e733742781259a1ee4b099cc3803559475
130781c130781
< Consensus hash for block 357778: 8be77e8ed18f904e12f9b06299591e390b9b095d428f2bd2612009383ef39d04

---
> Consensus hash for block 357778: c8554bd204e6418c81a680915cc4a7ec10c8a480b05c7f729208204874241ab1
135254c135254
< Consensus hash for block 361484: be507dfe1196e3dd48bad6428a2187ccb49e4fad8c8352a1aff7b5b527d4a672

---
> Consensus hash for block 361484: d49ac6171fb894c99f2bed22cfff51840446ffb0842337ce979343ce80f159d0
135903,135904c135903,135904
< Consensus hash for block 361999: 446c15064f4635e5fa6233eade38f5f0cfc178d4a8b1db20607033b260ad602f
< Consensus hash for block 362000: 446c15064f4635e5fa6233eade38f5f0cfc178d4a8b1db20607033b260ad602f

---
> Consensus hash for block 361999: 0ae8ab177298a60583dc88a6e1811238bbb6f69dc7ed794ce470eb8368696a20
> Consensus hash for block 362000: 0ae8ab177298a60583dc88a6e1811238bbb6f69dc7ed794ce470eb8368696a20
138910c138910
< Consensus hash for block 364443: 8d6cc9642619880f15b2f247b6537c9f102b30a8658faf420ba6f217667b6da4

---
> Consensus hash for block 364443: 427948865aa781e5f54bf497d92b02fc0e4411ca8e451a2d56b7ca26ab1e4e0d
141894c141894
< Consensus hash for block 366548: 72ef3b452c3712d839d7ee1ba5828f7c3d809c10ae9cb461f2c4fe0925e6448b

---
> Consensus hash for block 366548: e984f282459f6d5a471fd7ed281d2319db17ea5fa19cbc032862d41031477028
151692c151692
< Consensus hash for block 373642: 12c8582b829e8977cf84e56cd02090bf1e43c2b06745c59a618b2761b275330a

---
> Consensus hash for block 373642: 69c77ed64947ce9fca25a595d3acef10d77383eff41f81c52b84bad248076d58
155555c155555
< Consensus hash for block 375845: 6d3a15ad4726d4a651d0028c39bebb9c59f9a3d74c78f42e985bc178e8467a19

---
> Consensus hash for block 375845: 7ef2fb6248a24a2c2d9de3c6236fdf87f6d39c48a5ee0ccf03886bb1f0e8bc54
177241c177241
< Consensus hash for block 390319: 3b210a0e81ecbbd7d3129c6f5d2080ecf49532f0a277ed3659884c1065367b6c

---
> Consensus hash for block 390319: ccf1995b6a15248e8c0bfc289cdc685a279bb698b01f370acb52fc410996d1dd
177380c177380
< Consensus hash for block 390440: 6e47142d8c16705c5ca60a99747d02b88571ce4c9e335bcd2a7fc37c42f3f463

---
> Consensus hash for block 390440: 15122f1c0c86b2095cf6a5fb3335c94d301f188011af9e3f5931e87e0e940ad1
178216c178216
< Consensus hash for block 391066: 811f2b55ae70a9997a31a6e3aebd97630cca33161ba131542e780424cf92e81b

---
> Consensus hash for block 391066: 63da2eed3d0d2dfdc9e4a5775a6668b1ad15d8e20080a220490145f17584984a
179709c179709
< Consensus hash for block 391713: fbeaa8917abc952aabbd4e87593cb4c5828dd20c32df2194e749bbdaaed42625

---
> Consensus hash for block 391713: facdc49d1e55c790c554b2f535b4d82ee711d78bb90480574132376cb751a190
179745,179767c179745,179767
< Consensus hash for block 391736: 44c443465a823f927ed589e9ecb8abb464c9164ae4644f219e65bafeef580420
< Consensus hash for transaction 06fd1a93d21daad4815d110dce4bb5f1aef57094b377d556d848a8a9e4ad8474: b6c69e6de272ab678667ffd448d60681818bff5e6b4e16c18a6d5f7f88da510a
< Consensus hash for transaction 3896c717805a2854bbb5ea669ed426ddc1699125876fcd342d5126fe4c9a0c8c: 1d7c27e4157b6f1206a948f7b7ef67d2c668968ea6bf5e4bf21bdd0a8fc348ed
< Consensus hash for transaction 02ed07f18b21c55484ae8400f78bda8578151941ec1d55791392099db4e46555: bb3132fb1ad11c180fad2ea57e5663543fad5b10dc8866cdd3d5cd5662cb17f9
< Consensus hash for transaction e40a9289705145786aa830087eed634a418508e7d43cca3b8cf753dff9bd7684: fde89aa5ae9dcaa08ef3adf1d2cec8823f860018992923d51838dc5c24f5fe1c
< Consensus hash for transaction 7eb9108a3a563e1b994ae8ecf37b77b346294ad5010e4cdb164648ed1df2d66e: 6186d3a10916c664d3ac024d0cde77064308bcc232e869222652098ae26ebef0
< Consensus hash for transaction d382f0a28601c6bbccc19cc6822473095f6961fa4d04088d6d34438e36568fb3: 1c2a35f8ab40d3faaf0d7ca2c7c8bc497daf968bbfa107fc259c63c081b4c59c
< Consensus hash for transaction a6100a1c143edfa3a70f8cc9d4946d5ab8a3559f5dc9b557303a40987a5689be: 2fe25f20520cdb9c1fecf68066d88ead6201641143476a602391fa207a0a323c
< Consensus hash for transaction bb905723ad169b45d94f6dd645819b71c05e81987e3f1213f37107774253871c: 14639dbbeb92a60974e5656cccee8e138452b2cb8e0dc3361eb6eb757a80e3a9
< Consensus hash for transaction 1d5e9beb5f92deba629454453625633c1aec48c1a8cb45669b4db9f750762ff4: ef72cee06f9780cf734f053b93a96800c3e78c7724de4dc59d4876dd1f25238f
< Consensus hash for transaction b2f98c7b35c1d01480e0304e3336bbb8a05d724ff8ce468c49b5da6f8174e6b1: c467ee6fffbb1fb421d134490f8dad30f0a08605eafe81b7bef9fcb19d191e27
< Consensus hash for transaction 5146eb2adc8719ad86405bef8c68bebe0912a04e7c2bd6e7eba6df06375311ec: b8218fee3f6240d8deb46e068ca4964c4dabe60e5702efaf617006f931501833
< Consensus hash for transaction 9c6f43865d7d7c9eddea546899e040e8be57a286b9298cf856af5d99dd71f72a: f38e7819899c3f71c71ef9965904a0cdff01ae7b038e326de9a9a4c659c6965d
< Consensus hash for transaction 9b49b54494e823deb6bd9cd927219d4ea0ccba7fb1d03e84add7603d00c09136: 8ea86324bd795cbfa11c399756650ea030e842893d0ad2aa5732c8e2246c31d1
< Consensus hash for transaction dcc8b793d4e759af2a5c11232771f8e48d8995d5eeee4c9582895162aed0946c: 0e36be0a34e654c2a60570c0ed3359f552b734ae8404c00bc81a155108157344
< Consensus hash for transaction 6a9294a755c38706508246b470c84671f1d09ff00e5f3b7dce752e1f6dc4b9c8: cce7d1d09f46926ba65e6fdc35f3cdc4e098c0891af0e177fec10f8cd185dbdd
< Consensus hash for transaction 39595516dc28a918bdd862fa4d8f50d623601ac57470c6e01b7074bb4f02f424: ba76de04a693ec79cb431b047282455722be6ba274ac3d52a6106e1e9d304c0b
< Consensus hash for transaction 41f6817d21d6c25b543eefdf907aeaf7f94d661678caf0f402808be5f8fa4a03: caafbdf23e62c625ff8467e5d783c0df12f64074d057f5e4ce2a6d9e077c45c8
< Consensus hash for transaction 4ca74eaa085efef8501553ddd5fd8dfc2a209d0c6052821e03bf8d08e8641741: b5675f7df9e28dc1f14f345917cf9597148422a476a4e08fb8aa1d9dae63ac86
< Consensus hash for transaction 00b1f6f83af8f2b3d57e06dc9fb9fda6a945df55afb7e80ed7e799b578d555cd: b08323089f90f129fcfe654def9f75974fe2d00ba1fac47a8cce2bae93c14e22
< Consensus hash for transaction 3eb7f8d017dae19d5d1de47dbff7cd14e6f05ca9f4752556e423cdc8a1ec0610: 0f65fd1318c091955dbd667c76154e3abb173bb45159178d2bd59e1b8132329b
< Consensus hash for transaction 7afe8c43bcd4fca8694558f2d85454e18bd90cbb3d73f35504eaaa11825f6e88: f460f2c48f12d16cb9eb2ab58c015cc30f61dec0b9ff0d50e65f3e87ce26d179
< Consensus hash for transaction 98775a0ca5466e79016c23df3c891131c540ac70a51438b7847be58bba37c3b7: 2f0d9d68e0059541acb5772475708eacc3fde809e7e645eb6d2b6fd35449cc09

---
> Consensus hash for block 391736: 06ced527a43aa52218ecfcaf920c674d42b6d1e11a1356314ea2b6b44cc5cc11
> Consensus hash for transaction 06fd1a93d21daad4815d110dce4bb5f1aef57094b377d556d848a8a9e4ad8474: d807a340d9bb9e6b1a7084c214f155ee9b687590a64ce2639f096e22bbc2cb88
> Consensus hash for transaction 3896c717805a2854bbb5ea669ed426ddc1699125876fcd342d5126fe4c9a0c8c: 5898779e4afcf32d253028f0a51f9626ac607da822b872a3635d8f8a21be2194
> Consensus hash for transaction 02ed07f18b21c55484ae8400f78bda8578151941ec1d55791392099db4e46555: 2daf547667d31a7e419504215945519ff7d2032880bbe8c5e7a6f02b66d0758e
> Consensus hash for transaction e40a9289705145786aa830087eed634a418508e7d43cca3b8cf753dff9bd7684: 14c5b6528902115e296e78b5d3de80423bbbba9561e5f310de7b255faa1cc662
> Consensus hash for transaction 7eb9108a3a563e1b994ae8ecf37b77b346294ad5010e4cdb164648ed1df2d66e: 00464ada70aa603d00fbad7dae457070f86af28d5c5d033a9eeda5ef07ca744e
> Consensus hash for transaction d382f0a28601c6bbccc19cc6822473095f6961fa4d04088d6d34438e36568fb3: 62e13a6a3e0019395db5fe37ffb924101661a309be42be9688f33ba5c937773b
> Consensus hash for transaction a6100a1c143edfa3a70f8cc9d4946d5ab8a3559f5dc9b557303a40987a5689be: 47aa233ba79f8a9f7a8901d6236c4ff431910255499df70582832c667cdd9dce
> Consensus hash for transaction bb905723ad169b45d94f6dd645819b71c05e81987e3f1213f37107774253871c: c55d1400d1f153a422ffc1338199950aa675c5573f0105dcac94b5e078f5e626
> Consensus hash for transaction 1d5e9beb5f92deba629454453625633c1aec48c1a8cb45669b4db9f750762ff4: 3183d16f8a06dfe2a827d7a5bda416a9ea0d34019933402a2078b045c9c7afcb
> Consensus hash for transaction b2f98c7b35c1d01480e0304e3336bbb8a05d724ff8ce468c49b5da6f8174e6b1: c887c4ca76bc60264dfee0ad266a741d6192f7bca482abacab0361099e0fe8f1
> Consensus hash for transaction 5146eb2adc8719ad86405bef8c68bebe0912a04e7c2bd6e7eba6df06375311ec: 0c456370cc563494c9955ef0c890107c64afee2f530e5ad14b194d8164ee31f7
> Consensus hash for transaction 9c6f43865d7d7c9eddea546899e040e8be57a286b9298cf856af5d99dd71f72a: c5520c357c4392ddab0d66990114be247b1756035d38e69460589e6516a0689d
> Consensus hash for transaction 9b49b54494e823deb6bd9cd927219d4ea0ccba7fb1d03e84add7603d00c09136: b6eb908fdcd572a8b17fe849b98b313f2642571524896274e2da48b584be3a63
> Consensus hash for transaction dcc8b793d4e759af2a5c11232771f8e48d8995d5eeee4c9582895162aed0946c: 41193792d64d0093ee663420ef0953affc419f44a308818273f31bbeec1f8da0
> Consensus hash for transaction 6a9294a755c38706508246b470c84671f1d09ff00e5f3b7dce752e1f6dc4b9c8: 91b9fe5cb35521ceefc27e6860a6e7cb27f5cea424ba87754bd8eea31decb963
> Consensus hash for transaction 39595516dc28a918bdd862fa4d8f50d623601ac57470c6e01b7074bb4f02f424: dc7a0cc3788a78a6cd68678c789182c74924760e01b27e89df7f8903932e1f4e
> Consensus hash for transaction 41f6817d21d6c25b543eefdf907aeaf7f94d661678caf0f402808be5f8fa4a03: 9ed083503c51365c55836d16421089670be7b9819311e81467ccc6e3cb4f3bb8
> Consensus hash for transaction 4ca74eaa085efef8501553ddd5fd8dfc2a209d0c6052821e03bf8d08e8641741: 46651a4819ffe5aef32dad04fcac823b688b7d11fa27cb982a2e550373adfbe2
> Consensus hash for transaction 00b1f6f83af8f2b3d57e06dc9fb9fda6a945df55afb7e80ed7e799b578d555cd: 6e8dfe76dc10d28bd2c13cb6107b05d713ab659b0bd02696cdc2f06846c0e970
> Consensus hash for transaction 3eb7f8d017dae19d5d1de47dbff7cd14e6f05ca9f4752556e423cdc8a1ec0610: 92780fd638fa647f78e6fbdb53db2569ad607fee7046bd828757adae53333b6b
> Consensus hash for transaction 7afe8c43bcd4fca8694558f2d85454e18bd90cbb3d73f35504eaaa11825f6e88: 633bda3c294e6f822678b0755e5a356e6680c8c07f4f25ace06525701c9c49af
> Consensus hash for transaction 98775a0ca5466e79016c23df3c891131c540ac70a51438b7847be58bba37c3b7: 8c4eab76f304cb650cebec8bb1db1ffcd71207329d5a9c782fb151d8272aaadd
```

**Comment by dexX7** [comment]
Sorry, it's a bit unclear to me what "converges and reconverges" means in this context. I assume the consensus starts to differ at some point, but when considering the whole picture, everything adds up as it should?

**Comment by zathras-crypto** [comment]
Just that the hashes will match up to a point, then differ for a block (or two), then start matching again.

**Comment by zathras-crypto** [comment]
Waiting for my current parse to finish, after which will switch to ```omnidebug=all``` and see what's happening between these blocks - FYI there are differences after blocks which contained no transactions, meaning this is likely some form of expiration calculation.

**Comment by dexX7** [comment]
> FYI there are differences after blocks which contained no transactions, meaning this is likely some form of expiration calculation.

Or DevMSC?

**Comment by zathras-crypto** [comment]
That's also possible - I'm running an all debug pass now to find out what's happening at those blocks (as well as staging the chain to a native 32-bit Windows instance to verify win x32 results)...

**Comment by dexX7** [comment]
How did you notice it in the first place?

**Comment by zathras-crypto** [comment]
I was running through a full cross-check of platform/architecture for both main net and test net requested by Patrick - mainly to verify that the ~16,000 or so trades on the testnet MetaDEx are all processed the same across all platforms and architectures - I did mainnet first (generating consensus hashes for all transactions and blocks) and that's where I noticed that Linux 32-bit had some diffs.

The only state change I see for the first differing block I'm checking is dev MSC, so I'm running the 32-bit side now to get the dev MSC values for the same block and see if there is a difference.

**Comment by dexX7** [comment]
> The only state change I see for the first differing block I'm checking is dev MSC, so I'm running the 32-bit side now to get the dev MSC values for the same block and see if there is a difference.

Very interesting, and I'd be very glad, if it turns out to be an issue with DevMSC.

**Comment by zathras-crypto** [comment]
I think it is DevMSC :)

No consensus difference for block 264905, and dev MSC matches.
Consensus hash changes for block 264906, and dev MSC differs between x32 and x64:

```diff
zathras@zathras-desktop:~/temp_block_logs$ diff 264906_devmsc_32bit 264906_devmsc_64bit 
2,4c2,4
< devmsc=507030922980, exodus_prev=507014601528, exodus_delta=16321452
< update_tally_map(1EXoDusjGwvnjZUyKkxZ4UHEf77z6A5S4P, 1=0x1, +16321452, ttype=0): before=507014601528, after=507030922980
< devmsc for block 264906: 16321452, Exodus balance: 5070.30922980
---
> devmsc=507030922981, exodus_prev=507014601528, exodus_delta=16321453
> update_tally_map(1EXoDusjGwvnjZUyKkxZ4UHEf77z6A5S4P, 1=0x1, +16321453, ttype=0): before=507014601528, after=507030922981
> devmsc for block 264906: 16321453, Exodus balance: 5070.30922981
```

Then the next block (264907) the dev MSC difference is corrected and the consensus hash returns to normal:
```diff
zathras@zathras-desktop:~/temp_block_logs$ diff 264907_devmsc_32bit 264907_devmsc_64bit 
2,4c2,4
< devmsc=507097784113, exodus_prev=507030922980, exodus_delta=66861133
< update_tally_map(1EXoDusjGwvnjZUyKkxZ4UHEf77z6A5S4P, 1=0x1, +66861133, ttype=0): before=507030922980, after=507097784113
< devmsc for block 264907: 66861133, Exodus balance: 5070.97784113
---
> devmsc=507097784113, exodus_prev=507030922981, exodus_delta=66861132
> update_tally_map(1EXoDusjGwvnjZUyKkxZ4UHEf77z6A5S4P, 1=0x1, +66861132, ttype=0): before=507030922981, after=507097784113
> devmsc for block 264907: 66861132, Exodus balance: 5070.97784113
```

**Comment by dexX7** [comment]
Pew.. thanks for testing all this.

It's a rather wild guess, but I assume it probably comes down to be an issue with `double`, `pow`, `long double` or `roundl`.

Given that we now have input values (block timestamp, previous MSC etc.), it could be tackled isolated, e.g. in an unit test. I'm going to setup a VM for 32b tomorrow and do some testing.

**Comment by zathras-crypto** [comment]
Awesome thanks, I'm just running the testing again on a true x86 windows (not wow64) and I'll let you know if that makes a difference.

I'm also going to send that alert out to 0.0.9 clients a bit later once my 0.0.9 node has synced up (I want it synced to check behaviour).

**Comment by zathras-crypto** [comment]
Verified that this does not seem to affect Win 32-bit (native), only Linux 32-bit.

**Comment by dexX7** [comment]
Just a quick heads up: I'm on it, but I have a bit of trouble with my VM. For some reason the performance on this machine in that combination is horrible, and I'm basically progressing in slow motion.

What I'd like to try:

1. Refactor [calculate_and_update_devmsc](https://github.com/OmniLayer/omnicore/blob/20172a9a8ef7500a85566fe6515327cddc42c3a0/src/omnicore/omnicore.cpp#L424-L439) so that it becomes a pure function to calculate DevMSC based on `timestamp` and `exodus_prev` (which is provided)
2. Create a new unit test for the new function with `timestamp = 1382289200`, `prev = 507014601528`
3. Check, whether the results on 32b and 64b are different
4. Finally add more debug stuff to pin it down

**Comment by dexX7** [comment]
It looks like the issue comes from here:

```c++
#include <cmath>
#include <stdint.h>

uint64_t rounduint64(long double ld)
{
    return static_cast<uint64_t>(roundl(fabsl(ld)));
}
```

It creates different results, both in Win and Linux 32:

`507030922980.500000`* is rounded to `507030922980`, and not `507030922981` as on 64 bit.

\* The string representation might not be accurate.

See:

- https://travis-ci.org/OmniLayer/omnicore/jobs/102218673#L1701 32b
- https://travis-ci.org/OmniLayer/omnicore/jobs/102218670#L1598 64b

It would probably be best to get away from using floating points for the Exodus calculations altogether.

**Comment by zathras-crypto** [comment]
Nice find, though surprised you found the difference on both Win x32 and Linux x32, as I'm sure my Win x32 testing didn't reveal any differences...

Agreed on getting rid of float :)

**Comment by dexX7** [comment]
Now here is the very odd part, I was wrong about `rounduint64` as it seems.

Adding a print line removes the issue: https://github.com/dexX7/bitcoin/commit/a8a8cf1bce2a4605c9bbfeb9273e5df047af676b

Using plain integers for the Exodus calculations is still the way to go in my opinion, but I fear this could break consensus, so I'm trying to get a better understanding of the underlying issue. I'm kinda stuck at this moment though..

**Comment by dexX7** [comment]
Resolved with 0.2.
